### PR TITLE
LE plugin rewrite

### PR DESCRIPTION
**Pull request by user7**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

I've been working on LE/LX plugin for some time. I intended just to apply relocations, but ended up with almost a complete rewrite. The original code was unsalvageable IMO, a fix to all its issues would be almost as heavy as a rewrite and would be very convoluted. Anyway, git diff for librz/bin/format/le/le.c is useless, one needs to read the whole thing to review it. 

The list of notable fixes:

1. Relocations are now applied, so analysis is able to decode jump tables.
2. Virtual files are used for patching, imports now point to reloc-targets virtual file. 
3. The minimum amount of continuous sections is created. The old code used to create a section per memory page. This is frustrating when file has hundreds of pages, but just two LE objects (segments). Those objects are usually continuous and such file should have just two sections. However pages can occur out of order or some pages may be compressed, so it's not always possible to create a single continuous section per object.
4. DOS extender bound LE executables are supported now. The old code used to work only in pure LE or MZ+LE cases. This is for working with DOS-era software, a lot of games used DOS/4GW then. 
5. Compressed (type 5) pages are supported now.
6. Added some support for 16-bit relocations.

Besides numerous small issues were fixed, misparsing LE entries, phantom fixups, memory leaks, undefined behavior, endless loops, etc. The new code should be more robust on malformed input.

I am not sure if I should elaborate on the implementation and in what detail. Code has some comments, should be reasonably readable, but let me know if you have any questions.

...

**Test plan**

I've updated tests accordingly. I ran ```rz-test test/db/formats/le``` when testing. I also tested on multiple OS/2 warp 4.52 files, but those are probably copyrighted.

...

**Closing issues**

Is part of #1530

...

### CODE REVIEW

**Review by XVilka — CHANGES_REQUESTED** [review]
Will check for changes later, but is it possible to compile some hello-world with necessary properties for OS/2 Warp for creating such testcases? Maybe OpenWatcom can? http://ptspts.blogspot.com/2020/04/openwatcom-exeprog.html

**Comment by user7** [comment]
Contact @svr on rizin Mattermost for questions.

**Comment by user7** [comment]
> Will check for changes later, but is it possible to compile some hello-world with necessary properties for OS/2 Warp for creating such testcases? Maybe OpenWatcom can?

Yes, possible, but not easy, there are a lot of things to test. I am more worried about general commit, I am new to rizin and might have made some questionable decisions. I'd rather someone skim read through the code and point out obvious issues. It may affect how further tests are written.

File: librz/bin/format/le/le.c [hunk 1]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+static void le_import_free(LE_import *imp) {
+	if (imp) {
+		free(imp->proc_name);
+		free(imp);
+	}
+}
+
+static ut32 le_import_hash(LE_import *imp) {
+	ut32 ho = ((imp->mod_ord + 1) * (imp->proc_ord + 1) + 12345) * 1103515245;
+	return sdb_hash(imp->proc_name) ^ ho;
```
Comment (wargio) [inline librz/bin/format/le/le.c]: out of pure curiosity, is this random hash or is defined somewhere in the file format?

**Review by wargio — CHANGES_REQUESTED** [review]
could you rename the branch to `fuzz-le` ? there were multiple fuzz binaries that exploited issues on the old LE code.

File: librz/bin/format/le/le.c (hunk 1 above)
Comment (user7) [reply librz/bin/format/le/le.c]: A random hash. I couldn't find hash combining integers in rizin, and just slapped this one in. It has no relation to the format it's just for using hashtable.

**Comment by user7** [comment]
> could you rename the branch to `fuzz-le` ? there were multiple fuzz binaries that exploited issues on the old LE code.

won't it kill the pull request?

File: librz/bin/format/le/le.c [hunk 2]
```diff
@@ -1,10 +1,40 @@
...
+// SPDX-FileCopyrightText: 2023 svr <svr.work@protonmail.com>
 // SPDX-License-Identifier: LGPL-3.0-only
 
 #include "le.h"
 #include <rz_bin.h>
+#include <rz_types.h>
+#include <sdbht.h>
 
-const char *le_get_module_type(rz_bin_le_obj_t *bin) {
+// Sources:
```
Comment (XVilka) [inline librz/bin/format/le/le.c]: Please use Doxygen syntax e.g. `\file`, etc

File: librz/bin/format/le/le.c [hunk 3]
```diff
@@ -1,10 +1,40 @@
...
+
+#define CHECK_READ(X, tmp, out) \
+	CHECK(rz_buf_read##X##_offset(buf, offset, &tmp) && *offset <= offset_end) \
+	out = tmp;
+
+#define CHECK_READ8(out)  CHECK_READ(8, tmp8, out)
+#define CHECK_READ16(out) CHECK_READ(_le16, tmp16, out)
+#define CHECK_READ32(out) CHECK_READ(_le32, tmp32, out)
+
+// --- Auxilliary functions -----------------------------------------------------------------------
```
Comment (XVilka) [inline librz/bin/format/le/le.c]: Same, use Doxygen to split categories

File: librz/bin/format/le/le.c [hunk 4]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+
+	ut8 *str = calloc((size_t)len + 1, sizeof(char));
+	if (!str) {
+		return false;
+	}
+	if (!rz_buf_read_offset(buf, offset, str, len)) {
+		free(str);
+		return false;
+	}
+	for (ut8 *s = str; s != str + len; ++s) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c]: For strings we might have better functions, not sure if it's appropriate to use them here though. cc @wargio

File: librz/bin/format/le/le.c [hunk 5]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
 		return NULL;
 	}
-	(*offset)++;
-	char *str = calloc((ut64)size + 1, sizeof(char));
-	rz_buf_read_at(buf, *offset, (ut8 *)str, size);
-	*offset += size;
-	return str;
+	RzBinImport *import = RZ_NEW0(RzBinImport);
+	if (!import) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:204]: Is this goto label unnecessary?

File: librz/bin/format/le/le.c [hunk 6]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
-	default:
-		memset(e, 0, sizeof(LE_entry_bundle_entry));
+static RZ_BORROW LE_import *le_add_import(rz_bin_le_obj_t *bin,
+	ut32 mod_ord, bool proc_by_ord, ut32 proc, ut32 sym_ord) {
+
+	RzBinImport *bin_imp = NULL;
+	LE_import *le_imp = NULL;
+	char *proc_name = NULL;
+	if (false) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:255]: The whole block unnecessary?

File: librz/bin/format/le/le.c [hunk 7]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+static bool page_write(rz_bin_le_obj_t *bin, LE_page *io_page, ut32 vaddr, ut8 *buf, ut32 len) {
+	return page_io(bin, io_page, vaddr, buf, len, false);
+}
+
+static ut32 le_reloc_vaddr(rz_bin_le_obj_t *bin, LE_reloc *reloc) {
+	return bin->le_pages[reloc->src_page].vaddr + reloc->src_off;
+}
+
+static ut32 le_reloc_size(LE_reloc *reloc) {
+	static const ut8 szmap[] = {
```
Comment (XVilka) [inline librz/bin/format/le/le.c:407]: Probably better to extract to the top level

File: librz/bin/format/le/le.c [hunk 8]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+		[FIXUP_OFF16] = 2,
+		[FIXUP_OFF32] = 4,
+		[FIXUP_SEL16_OFF16] = 4,
+		[FIXUP_REL32] = 4,
+		[FIXUP_SEL16_OFF32] = 6,
+	};
+	return reloc->type < sizeof(szmap) ? szmap[reloc->type] : 0;
+}
+
+// --- Loading all things LE from the binary ------------------------------------------------------
```
Comment (XVilka) [inline librz/bin/format/le/le.c]: Doxygen category/paragraph

File: librz/bin/format/le/le.c [hunk 9]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+}
+
+// Loading objects, see [1] 3.4 Object Table
+static bool le_load_objects(rz_bin_le_obj_t *bin) {
+	LE_header *h = bin->header;
+	LE_object *objects = NULL;
+	if (!h->objcnt) {
+		return true; // no objects, binary is a forwarders-only library
+	}
+	if (false) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c]: Unnecessary block

File: librz/bin/format/le/le.c [hunk 10]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	}
+	bin->objects = objects;
+	return true;
+}
+
+// Loading a single non-empty entry, see [1] 3.8.1-4
+static bool le_load_entry_record(rz_bin_le_obj_t *bin, ut64 *offset, ut8 type, ut32 obj_num,
+	RzVector /*<LE_entry>*/ *entries) {
+
+	if (false) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c:580]: Unnecessary

File: librz/bin/format/le/le.c [hunk 11]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
-RzList /*<char *>*/ *le_get_entries(rz_bin_le_obj_t *bin) {
-	ut64 offset = (ut64)bin->header->enttab + bin->headerOff;
-	RzList *l = rz_list_newf(free);
-	if (!l) {
+// Loading symbols, see [1] 3.8 Entry Table
+static RZ_OWN RzVector /*<LE_entry>*/ *le_load_entries(rz_bin_le_obj_t *bin) {
+	char *name = NULL;
+	RzVector *entries = rz_vector_new(sizeof(LE_entry), NULL, NULL);
+	if (!entries) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:656]: Unnecessary goto label

File: librz/bin/format/le/le.c [hunk 12]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+
+	return entries;
+}
+
+// Loading page map, see [1] 3.5 Object Page Table
+static RZ_OWN LE_page *le_load_pages(rz_bin_le_obj_t *bin) {
+	LE_header *h = bin->header;
+	LE_page *le_pages = calloc(h->mpages, sizeof(LE_header));
+	if (!le_pages) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:748]: Unnecessary label

File: librz/bin/format/le/le.c [hunk 13]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	return le_pages;
 }
 
-RzList /*<RzBinImport *>*/ *rz_bin_le_get_imports(rz_bin_le_obj_t *bin) {
-	RzList *l = rz_list_newf((RzListFree)rz_bin_import_free);
-	if (!l) {
-		return NULL;
+// See [2] UnpackMethod1 for the algortihm.
+static void le_unpack_iterated(rz_bin_le_obj_t *bin, ut8 *out, ut32 out_size, LE_page *page) {
+	if (false) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c:852]: Unnecessary block

File: librz/bin/format/le/le.c [hunk 14]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+// See [2] UnpackMethod2 for the algortihm.
+static void le_unpack_compressed(rz_bin_le_obj_t *bin, ut8 *out, ut32 out_size, LE_page *page) {
+	ut8 tmp8;
+	ut16 tmp16;
+	RzBuffer *buf = bin->buf;
+	ut64 off = page->paddr, *offset = &off;
+	ut64 offset_end = *offset + page->psize;
+	ut32 out_size_total = out_size;
+
+	if (false) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c:895]: Same

File: librz/bin/format/le/le.c [hunk 15]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
 
-	offset += sizeof(ut16);
-	ut16 data_size;
-	if (!rz_buf_read_ble16_at(bin->buf, offset, &data_size, h->worder)) {
-		return;
+static RzVector /*<LE_map>*/ *le_create_maps(rz_bin_le_obj_t *bin) {
+	RzVector *le_maps = rz_vector_new(sizeof(LE_map), le_map_fini, NULL);
+	ut8 *tmp_buf = NULL;
+	if (!le_maps) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1007]: Same

File: librz/bin/format/le/le.c [hunk 16]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
 }
 
-char *le_get_modname_by_ord(rz_bin_le_obj_t *bin, ut32 ordinal) {
+static RZ_OWN RzPVector /*<char*>*/ *le_load_import_mod_names(rz_bin_le_obj_t *bin) {
+	RzPVector *names = rz_pvector_new(free);
 	char *modname = NULL;
-	ut64 off = (ut64)bin->header->impmod + bin->headerOff;
-	while (ordinal > 0) {
+	if (!names) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1141]: .

File: librz/bin/format/le/le.c [hunk 17]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
-	if (!rz_buf_read_ble32_at(bin->buf, (ut64)h->fpagetab + bin->headerOff + cur_page * sizeof(ut32), &tmp_offset, h->worder)) {
-		rz_list_free(l);
-		rz_list_free(entries);
-		rz_list_free(sections);
-		return NULL;
+static bool le_patch_relocs(rz_bin_le_obj_t *bin) {
+	// possibly resize vfiles
+	ut32 *max_vaddr = NULL;
+	LE_map **last_map = NULL;
+	if (false) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1161]: .

File: librz/bin/format/le/le.c [hunk 18]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	}
+	return true;
+}
+
+static bool le_load_fixup_record(rz_bin_le_obj_t *bin, RzList /*<RzBinReloc *>*/ *relocs_out,
+	ut32 page_i, ut64 *offset, ut64 offset_end) {
+
+	LE_header *h = bin->header;
+	ut64 start_offset = *offset;
+	if (false) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1305]: .

File: librz/bin/format/le/le.c [hunk 19]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	ht_pp_free(bin->le_import_ht);
+	rz_list_free(bin->le_relocs);
+	free(bin);
+}
+
+void rz_bin_le_destroy(RzBinFile *bf) {
+	rz_bin_le_free(bf->o->bin_obj);
+}
+
+RZ_OWN bool rz_bin_le_load_buffer(RzBinFile *bf, RzBinObject *obj, RzBuffer *buf, Sdb *sdb) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c]: bool doesn't need ownership

File: librz/bin/format/le/le.c [hunk 20]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	rz_return_val_if_fail(bf && obj && buf, false);
 	rz_bin_le_obj_t *bin = RZ_NEW0(rz_bin_le_obj_t);
+	char const *err_ctx = ", unable to load file header.";
 	if (!bin) {
-		return NULL;
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1527]: unused label?

File: librz/bin/format/le/le.c [hunk 21]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	}
+	return rz_list_clone(bin->symbols);
+}
+
+RZ_OWN RzList /*<RzBinSection *>*/ *rz_bin_le_get_sections(RzBinFile *bf) {
+	rz_bin_le_obj_t *bin = bf->o->bin_obj;
+	RzList *sections = rz_list_newf((RzListFree)rz_bin_section_free);
+	RzBinSection *sec = NULL;
+	if (!sections) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1597]: same

File: librz/bin/format/le/le.c [hunk 22]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	return sections;
+}
+
+RZ_OWN RzList /*<RzBinAddr *>*/ *rz_bin_le_get_entry_points(RzBinFile *bf) {
+	rz_bin_le_obj_t *bin = bf->o->bin_obj;
+	LE_header *h = bin->header;
+	RzBinAddr *addr = NULL;
+	RzList *entries = rz_list_newf((RzListFree)free);
+	if (!entries) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1639]: same

File: librz/bin/format/le/le.c [hunk 23]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
-	free(bin->filename);
-	free(bin);
+RZ_OWN RzList /*<char *>*/ *rz_bin_le_get_libs(RzBinFile *bf) {
+	rz_bin_le_obj_t *bin = bf->o->bin_obj;
+	if (rz_pvector_empty(bin->imp_mod_names)) {
+		return NULL;
+	}
+	RzList *libs = rz_list_new();
+	if (!libs) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1677]: .

File: librz/bin/format/le/le.c [hunk 24]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+#define VFILE_NAME_PATCHED       "patched"
+#define VFILE_NAME_RELOC_TARGETS "reloc-targets"
+
+RzList /*<RzBinVirtualFile *>*/ *rz_bin_le_get_virtual_files(RzBinFile *bf) {
+	rz_bin_le_obj_t *bin = bf->o->bin_obj;
+	RzBinVirtualFile *vf = NULL;
+	RzBuffer *buf = NULL;
+	RzList *vfiles = rz_list_newf((RzListFree)rz_bin_virtual_file_free);
+	if (!vfiles) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1697]: .

File: librz/bin/format/le/le.c [hunk 25]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	return vfiles;
+}
+
+RZ_OWN RzList /*<RzBinReloc *>*/ *rz_bin_le_get_relocs(RzBinFile *bf) {
+	rz_bin_le_obj_t *bin = bf->o->bin_obj;
+	RzList /*<LE_reloc *>*/ *le_relocs = bin->le_relocs;
+	RzList /*<RzBinReloc *>*/ *relocs = rz_list_newf((RzListFree)rz_bin_reloc_free);
+	RzBinReloc *reloc = NULL;
+	if (!relocs) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1748]: .

File: librz/bin/format/le/le.c [hunk 26]
```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+	}
+	return relocs;
+}
+
+RZ_OWN RzList /*<RzBinMap *>*/ *rz_bin_le_get_maps(RzBinFile *bf) {
+	rz_bin_le_obj_t *bin = bf->o->bin_obj;
+	RzList *maps = rz_list_newf((RzListFree)rz_bin_map_free);
+	RzBinMap *map = NULL;
+	if (!maps) {
+	fail_cleanup:
```
Comment (XVilka) [inline librz/bin/format/le/le.c:1809]: .

File: librz/bin/p/bin_le.c [hunk 27]
```diff
@@ -4,149 +4,114 @@
...
-
-static bool load_buffer(RzBinFile *bf, RzBinObject *obj, RzBuffer *buf, Sdb *sdb) {
-	rz_return_val_if_fail(bf && obj && buf, false);
-	rz_bin_le_obj_t *res = rz_bin_le_new_buf(buf);
-	if (res) {
-		obj->bin_obj = res;
-		return true;
+static RzBinInfo *le_info(RzBinFile *bf) {
+	RzBinInfo *info = RZ_NEW0(RzBinInfo);
+	if (info) {
```
Comment (XVilka) [inline librz/bin/p/bin_le.c]: reverse condition for early exit

File: test/db/formats/le [hunk 28]
```diff
@@ -381,41 +343,22 @@ aac
...
-BROKEN=1
 FILE=bins/le/cdogs.exe
 CMDS=pi 1
 EXPECT=<<EOF
-jmp 0x465
-EOF
-RUN
-
-NAME=LE: cdogs.exe segments
-BROKEN=1
```
Comment (XVilka) [inline test/db/formats/le:407]: Please also list segments/sections

**Review by XVilka — CHANGES_REQUESTED** [review]
Please be careful so that it doesn't break Travis CI System Z worker (big endian host).

File: librz/bin/format/le/le.c [hunk 29]
```diff
@@ -57,828 +87,1749 @@
...
+			ut32 record;
+			CHECK(rz_buf_read_be32_offset(bin->buf, &offset, &record));
+			ut32 page_num = record >> 8;
+			ut8 page_flags = record & 0xFF;
+			if (page_flags != 0) {
+				RZ_LOG_WARN("LE: unsupported LE page flags 0x%02x for page #%d.\n",
+					page_flags, page_i + 1);
+			}
+			page->type = PAGE_LEGAL;
+			page->paddr = bin->mz_off + h->datapage + (page_num - 1) * h->pagesize;
```
[the remark this answers is not in the document]
Comment (wargio) [reply librz/bin/format/le/le.c]: please use the overflow macro checks before doing this math.

File: librz/bin/format/le/le.c (hunk 4 above)
Comment (wargio) [reply librz/bin/format/le/le.c]: depends if this is only used here or elsewhere.

File: librz/bin/format/le/le.c (hunk 5 above)
Comment (user7) [reply librz/bin/format/le/le.c:204]: it's used by CHECK macros below

File: librz/bin/format/le/le.c (hunk 6 above)
Comment (user7) [reply librz/bin/format/le/le.c:255]: same, used by CHECK macros, such blocks are all over the code and used the same way

File: librz/bin/format/le/le.c (hunk 7 above)
Comment (user7) [reply librz/bin/format/le/le.c:407]: I don't export exotic relocs from rz_bin_le_get_relocs, for example 6 byte 16:32 reloc is converted to a pair of a 32 bit reloc and a 16 bit selector reloc. I would need to add a new reloc type to do that properly, and I am not up to the task. So I judged it's an LE internal detail. I think it should not be exported.

File: librz/bin/format/le/le.c (hunk 9 above)
Comment (user7) [reply librz/bin/format/le/le.c]: Which one? objcnt is needed to make debug build calloc happy when there's no objects. false is used by multiple CHECKs below.

File: librz/bin/format/le/le.c (hunk 19 above)
Comment (user7) [reply librz/bin/format/le/le.c]: fixed

File: test/db/formats/le (hunk 28 above)
Comment (user7) [reply test/db/formats/le:407]: There's a test "cdogs.exe sections". Regarding segments I have trouble representing them using rizin RzBin* structures. I'll explain in more detail when I have time. Currently there are no segments.

File: librz/bin/format/le/le.c (hunk 6 above)
Comment (XVilka) [reply librz/bin/format/le/le.c:255]: but it has `if (false)`

File: librz/bin/format/le/le.c (hunk 2 above)
Comment (user7) [reply librz/bin/format/le/le.c]: done

File: librz/bin/format/le/le.c (hunk 6 above)
Comment (user7) [reply librz/bin/format/le/le.c:255]: Because this particular block is not entered except by goto. Cleanup blocks are poor man's destructors. When used, function starts with declarations of variables to be destroyed in case of error, then follows the cleanup block. When those variables are initialized with a fallible function e.g. ```x = malloc()```, it makes sense to enter the cleanup block right away. In such case we enter under a meaningful condition, e.g. ```if (!x)```. Other times variables are just initialized by NULLs and there's no way for the error to occur before the cleanup block. In such cases we never enter it except by goto, so I put ```if (false)``` as enter condition. I could avoid using ```if (false)``` in a number of ways by rearranging code. However I prefer a uniform style, with the cleanup block always appearing at the top.

Granted some functions have only one or two CHECKs, and might be simplified by removing the cleanup block altogether. This is because I've been simplifying functions when coding and some lost most of their CHECKs. I strive to keep only those cleanup blocks that spare me at least some code duplication, but perhaps in some places both cleanup block and CHECKs can be dropped.

Also I find such blocks much easier to work with when reviewing and refactoring. Instead of remembering to delete the pointer in this or that case, I just delete them all in the cleanup section. And the list of affected pointers is plain to see at the top of the function.

File: librz/bin/p/bin_le.c (hunk 27 above)
Comment (user7) [reply librz/bin/p/bin_le.c]: done

File: librz/bin/format/le/le.c (hunk 1 above)
Comment (wargio) [reply librz/bin/format/le/le.c]: awesome! just as i thought.
Instead of using multiplication, use xor.
like `ut32 ho = ((imp->mod_ord + 1) * (imp->proc_ord + 1) + 12345) ^ 1103515245;`

Comment (wargio) [reply librz/bin/format/le/le.c]: because if you multiply by big number you can overflow the variable.

File: librz/bin/format/le/le.c (hunk 6 above)
Comment (user7) [reply librz/bin/format/le/le.c:255]: btw CHECK macro is what does ```goto fail_cleanup``` when condition is zero

File: librz/bin/format/le/le.c (hunk 3 above)
Comment (user7) [reply librz/bin/format/le/le.c]: Do you know the syntax by chance?

File: librz/bin/format/le/le.c (hunk 24 above)
Comment (user7) [reply librz/bin/format/le/le.c:1697]: CHECK macro does goto fail_cleanup on failure

File: librz/bin/format/le/le.c (hunk 26 above)
Comment (user7) [reply librz/bin/format/le/le.c:1809]: CHECK macro does goto fail_cleanup on failure

File: librz/bin/format/le/le.c (hunk 3 above)
Comment (XVilka) [reply librz/bin/format/le/le.c]: See https://www.doxygen.nl/manual/commands.html

Comment (user7) [reply librz/bin/format/le/le.c]: Never mind, just made it a doxygen comment.

File: librz/bin/format/le/le.c (hunk 1 above)
Comment (user7) [reply librz/bin/format/le/le.c]: But that is the intention, it is supposed to wrap around and keep the remainder modulo 2^32. I took generator from here https://en.wikipedia.org/wiki/Linear_congruential_generator#Parameters_in_common_use see line 5 of the table.

File: librz/bin/format/le/le.c (hunk 25 above)
Comment (user7) [reply librz/bin/format/le/le.c:1748]: CHECK macro does goto fail_cleanup on failure

File: test/db/formats/le (hunk 28 above)
Comment (user7) [reply test/db/formats/le:407]: On the second thought, there's no explanation needed. We have no segments now, only sections. So segment test has been removed.

**Comment by wargio** [comment]
regarding the fuzz tests, just run them in your machine to be sure it won't break the CI once merged.

File: librz/bin/format/le/le.c [hunk 30]
```diff
@@ -57,828 +103,1754 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
...
+			}
+		}
 	}
+	return 0;
+}
 
-	size &= 0x7F; // Max is 127
-	if (!size) {
+static void le_import_free(LE_import *imp) {
+	if (imp) {
```
Comment (XVilka) [inline librz/bin/format/le/le.c]: Reverse this condition

**Review by XVilka — CHANGES_REQUESTED** [review]
LGTM with minor nitpick. Please do not merge, we will use internal branch to trigger fuzzing and Travis CI (BE) jobs.

**Comment by user7** [comment]
A tests "Mixed linter and checks / licenses (pull_request)" says "not compliant with REUSE spec". It looks like the test itself is broken. Is it ok?

**Comment by XVilka** [comment]
> A tests "Mixed linter and checks / licenses (pull_request)" says "not compliant with REUSE spec". It looks like the test itself is broken. Is it ok?

You can just rebase on top of the `dev` where it's fixed already: https://github.com/rizinorg/rizin/commit/8f1710a403ecfdc14047f1854cbdaa6dfa49c13b

File: librz/bin/format/le/le.c (hunk 30 above)
Comment (user7) [reply librz/bin/format/le/le.c]: done

**Comment by user7** [comment]
May I squash small commits while rebasing?

**Comment by wargio** [comment]
> May I squash small commits while rebasing?

sure ! you can squash in one commit even

**Comment by user7** [comment]
Fuzz tests passed without errors.

**Comment by user7** [comment]
I am going to add couple of tests for more obscure features. Any specific requests what tests to add?

**Comment by XVilka** [comment]
> I am going to add couple of tests for more obscure features. Any specific requests what tests to add?

No specific requests. Just try to cover more code paths with these tests, or some edge cases. Please also rebase on top of the latest `dev`.

**Review by XVilka — CHANGES_REQUESTED** [review]
LGTM now, but I saw you added LX binary but no test script here.

**Comment by user7** [comment]
> LGTM now, but I saw you added LX binary but no test script here.

Yes, I had not committed it yet, I'll post here when ready. Saturday night likely.

**Comment by user7** [comment]
Added tests for new binaries, squashed everything again, I am good to merge after tests.

**Comment by XVilka** [comment]
Me or @wargio will import and open a new PR to check also on ARM and BE platforms later today. Thank you!

**Comment by wargio** [comment]
https://github.com/rizinorg/rizin/pull/3686

**Comment by XVilka** [comment]
It passed all tests, thank you!
